src/traces/bar/plot.js

```diff
@@ -108,12 +125,430 @@ module.exports = function plot(gd, plotinfo, cdbar) {
+        }
+    }
+    else {
+        if(y1 > y0) {
+            // bar end is on the bottom
+            targetX = (x0 + x1) / 2;
+            targetY = y1 + textpad + targetHeight / 2;
+            if(rotate) rotate = -90;  // rotate counter-clockwise
+        }
+        else {
+            targetX = (x0 + x1) / 2;
+            targetY = y1 - textpad - targetHeight / 2;
+            if(rotate) rotate = 90;  // rotate clockwise
+        }
+    }
+
+    return getTransform(textX, textY, targetX, targetY, scale, rotate);
+}
+
+function getTransform(textX, textY, targetX, targetY, scale, rotate) {
+    var transformScale,
+        transformRotate,
+        transformTranslate;
+
+    if(scale < 1) transformScale = 'scale(' + scale + ') ';
+    else {
+        scale = 1;
+        transformScale = '';
+    }
+
+    transformRotate = (rotate) ?
+        'rotate(' + rotate + ' ' + textX + ' ' + textY + ') ' : '';
+
+    // Note that scaling also affects the center of the text box
+    var translateX = (targetX - scale * textX),
+        translateY = (targetY - scale * textY);
+    transformTranslate = 'translate(' + translateX + ' ' + translateY + ')';
+
+    return transformTranslate + transformScale + transformRotate;
+}
+
+function getText(trace, index) {
+    var value = getValue(trace.text, index);
+    return coerceString(attributeText, value);
+}
+
+function getTextPosition(trace, index) {
+    var value = getValue(trace.textposition, index);
+    return coerceEnumerated(attributeTextPosition, value);
+}
+
+function getTextFont(trace, index, defaultValue) {
+    return getFontValue(
+        attributeTextFont, trace.textfont, index, defaultValue);
+}
+
+function getInsideTextFont(trace, index, defaultValue) {
+    return getFontValue(
+        attributeInsideTextFont, trace.insidetextfont, index, defaultValue);
+}
+
+function getOutsideTextFont(trace, index, defaultValue) {
+    return getFontValue(
+        attributeOutsideTextFont, trace.outsidetextfont, index, defaultValue);
+}
+
+function getFontValue(attributeDefinition, attributeValue, index, defaultValue) {
+    attributeValue = attributeValue || {};
+
+    var familyValue = getValue(attributeValue.family, index),
+        sizeValue = getValue(attributeValue.size, index),
+        colorValue = getValue(attributeValue.color, index);
+
+    return {
+        family: coerceString(
+            attributeDefinition.family, familyValue, defaultValue.family),
+        size: coerceNumber(
+            attributeDefinition.size, sizeValue, defaultValue.size),
+        color: coerceColor(
+            attributeDefinition.color, colorValue, defaultValue.color)
+    };
+}
+
+function getValue(arrayOrScalar, index) {
+    var value;
+    if(!Array.isArray(arrayOrScalar)) value = arrayOrScalar;
+    else if(index < arrayOrScalar.length) value = arrayOrScalar[index];
+    return value;
+}
+
+function coerceString(attributeDefinition, value, defaultValue) {
```

Comment:
Can you simply reuse [`Lib.coerce`](https://github.com/plotly/plotly.js/blob/master/src/lib/coerce.js) here?